### ISSUE TITLE
Compatibility with Boom2?

### ISSUE THREAD

**Issue by NobleShit** [kyleneideck/BackgroundMusic#71]
How would I get this to work best with Boom2? Run an Aggregate device?

**Comment by kyleneideck** [comment]
That was my first thought, too, but I think you'd run into a problem where both Boom2 and Background Music played your audio through your real audio device. You also wouldn't be able to control your system volume (or mute) normally because aggregate devices don't have controls.

You might be able to disable playthrough in Background Music so that only Boom2 plays audio. That would break per-app volumes but I think auto-pausing music would still work. You'd have to edit the code, though. I think you should be able to just comment out everything in BGMPlayThrough::start and BGMPlayThrough::WaitForOutputDeviceToStart (except the final return statements). Let me know if you try it and doesn't work and I'll look into it properly.

The other thing that might work is to just set the output device in Background Music to Boom2's virtual device, or vice versa. You'll probably want to try both ways around to see which breaks more functionality.

Keep in mind that using both at once will probably use a fair bit of CPU and increase your latency. Look at the CPU use of `coreaudiod` as well as the Background Music and Boom2 processes. I use [Stepmania](http://www.stepmania.com) to test latency because I had it installed already and that seems to work well.

Also, according to @ZV95, Background Music and Boom2 aren't compatible at all: #21. I don't have a copy of Boom2 to figure out why exactly, but that doesn't surprise me. Intercepting system audio isn't supported by CoreAudio, so both apps are using the CoreAudio API in a way it wasn't really designed for.

**Comment by NobleShit** [comment]
With both running, I get no audio at all. The main functionality I was looking fro was per-app volume. I have noticed that I can get audio to resume if I go into settings and chose “Headphone” as main output, at which point it bounces rick back to Boom (as it should, and normally does) but it still breaks per-app volume.

**Comment by kyleneideck** [comment]
Is "Headphone" the name Boom uses for its virtual audio device? (That could make sense, I just want to make sure.) And you're setting Background Music's output device to that virtual device?

Is Background Music Device set as your system's default device when you test this? For per-app volume to work it would have to be. But that would break any per-app features Boom has. And Boom might change your default device back, which would break things again.

**Comment by NobleShit** [comment]
No, Boom uses “Boom” Headphone is the built in output when I have a 3.5mm jacked plugged in (whether it be actual headphones or desktop speakers.) Would I want to run Boom before Background Music or vice versa? Boom has no per-app options, only system wide. If you would like, I have a cracked version of Boom2 (1.3) I can email you the .dmg.

**Comment by NobleShit** [comment]
Also a big concern of mine is with FaceTime(.app FWIW) I dont like how Apple handles it. I have tried a couple “mixers” (including Background Music) and none do anything to FaceTime and within a call, it quiets all other audio from anywhere else.

**Comment by kyleneideck** [comment]
>No, Boom uses “Boom” Headphone is the built in output when I have a 3.5mm jacked plugged in (whether it be actual headphones or desktop speakers.)

You'd need to have the "Boom" device selected as the output device in Background Music. Otherwise Boom wouldn't be able to get any audio.

>Would I want to run Boom before Background Music or vice versa

Probably Boom first to get Background Music's virtual device set as the system default device, but it depends on what Boom does. Check in System Preferences to be sure.

I'll try installing the trial version of Boom 2 when I get a chance and see if I can find out what's actually going on.

**Comment by kyleneideck** [comment]
I couldn't figure out a way to get per-app volume working with Boom 2 running at the same time. Boom seems to make sure Boom2Device is set as the default device whenever Boom is running. Short of asking Boom's developers to add an option to disable that, I don't see a way around it.

**Comment by NobleShit** [comment]
Well shoot. Thanks for all the help anyways! Much appreciated!

**Comment by sammarcus** [comment]
Hi @kyleneideck , was just curious if anything had changed on this in the last few years? I'm most interested in the pause functionality, does that at least work even if the per app volume does not? Thank you!

**Comment by sammarcus** [comment]
hi @kyleneideck 👋🏼🙂👀

**Comment by kyleneideck** [comment]
@sammarcus Nothing that I'm aware of, but I haven't looked into it since this issue was first opened. I think you'd need BGMDevice set as your system default device for auto-pausing to work, so unless Boom has changed something, I wouldn't expect it to.

In most cases, I probably wouldn't recommend using two apps that intercept the system audio at the same time anyway. macOS's audio system isn't designed to support them, even if you're only using one, so they have to work around it with virtual devices and that has performance costs.
